### PR TITLE
Add Tauri auto-updates

### LINKED ISSUE #365: [Tauri] Implement auto updates
See https://tauri.studio/en/docs/usage/guides/updater

### PR DESCRIPTION
**Pull request by BrunoBernardino**
Closes #365

Still unfinished, I need to learn where are the build signatures stored.

### CODE REVIEW

**Comment by MaKleSoft** [comment]
> I need to learn where are the build signatures stored.
According to https://tauri.studio/en/docs/usage/guides/updater#update-file-json-format:

```json
{
    "name":"v1.0.0",
    "notes":"Test version",
    "pub_date":"2020-06-22T19:25:57Z",
    "platforms": {
        "darwin": {
            "signature":"", <-- Signature goes here
            "url":"https://github.com/lemarier/tauri-test/releases/download/v1.0.0/app.app.tar.gz"
        },
        "linux": {
            "signature":"", <-- Signature goes here
            "url":"https://github.com/lemarier/tauri-test/releases/download/v1.0.0/app.AppImage.tar.gz"
        },
        "win64": {
            "signature":"", <-- Signature goes here
            "url":"https://github.com/lemarier/tauri-test/releases/download/v1.0.0/app.x64.msi.zip"
        }
    }
}
```

**Comment by BrunoBernardino** [comment]
This is now ready for review. The tag is at https://github.com/padloc/padloc/releases/tag/tauri-v4.0.0 and the release at https://github.com/padloc/padloc/releases/tag/untagged-c25cf3a6f57f16f96e85 though we should manually delete those before we merge, to avoid confusion.

File: .github/workflows/build-cordova.yml
Comment (BrunoBernardino) [inline .github/workflows/build-cordova.yml:8]: This is just for consistency, once this shows for manual triggering.

File: .github/workflows/build-tauri.yml
Comment (BrunoBernardino) [inline .github/workflows/build-tauri.yml:59]: This signs the builds, now.

File: .github/workflows/release-tauri.yml
Comment (BrunoBernardino) [inline .github/workflows/release-tauri.yml:1]: This works and will create a separate tag and release for tauri, for now. We can, instead, simply eventually rename this action, in a follow-up PR, to "Release", and have it create the release and upload everything it can (electron, cordova, web extensions, etc.), automatically, and then we just need to "finalize it" (publish as not pre-release in GitHub), then download and upload the files to the respective stores.

File: .github/workflows/release-tauri.yml
Comment (BrunoBernardino) [inline .github/workflows/release-tauri.yml:13]: We can try to deduce this from the files, but it's not a bad idea to manually set it, since we'll be creating a tag from this. At least for the first release or so.

File: packages/tauri/src-tauri/tauri.conf.json
Comment (BrunoBernardino) [inline packages/tauri/src-tauri/tauri.conf.json:61]: This URL won't really exist until we publish this tag. We'll also need to manually tag something as latest (which we can do in a follow-up "release" action, which can be the existing one, renamed, which would delete the current latest and tag the new code as such). We might want to change this to `latest` instead of `tauri-latest` if we don't want to consider those separate releases.

File: packages/tauri/src-tauri/Cargo.toml
Comment (MaKleSoft) [inline packages/tauri/src-tauri/Cargo.toml:3]: I don't think we need to bother updating this field, since the version from `tauri.conf.json` take precedence over this one.

File: packages/tauri/src-tauri/Cargo.toml
Comment (MaKleSoft) [inline packages/tauri/src-tauri/Cargo.toml:2]: Let's change this to lowercase, to get rid of that warning during the build process.

File: packages/tauri/src-tauri/tauri.conf.json
Comment (MaKleSoft) [reply packages/tauri/src-tauri/tauri.conf.json:61]: Does this mean we have to publish a separate tag named `tauri-latest` (or simply `latest`), in addition to the regular releases? What is the advantage over simply using `https://github.com/padloc/padloc/releases/latest/download/tauri-update.json`, which automatically resolves to the latest release?

**Review by MaKleSoft — COMMENTED** [review]
Great stuff! Have you been able to test this yet?

File: .github/workflows/release-tauri.yml
Comment (BrunoBernardino) [reply .github/workflows/release-tauri.yml:13]: It's not common practice, and I think a safer approach would be to confirm the versions in the code match what's being asked instead, but we can definitely have it update the files and commit, if that's easier.

File: packages/tauri/src-tauri/tauri.conf.json
Comment (BrunoBernardino) [reply packages/tauri/src-tauri/tauri.conf.json:61]: If we keep it as is, yes. Since this is not really part of "latest" right now, I kept it separate, but if we want to move forward with this (and keep Tauri as a target like Electron, for the same version), then we don't need this tag separation at all.

File: packages/tauri/generate-tauri-update-file.js
Comment (BrunoBernardino) [reply packages/tauri/generate-tauri-update-file.js:29]: :facepalm:

File: packages/tauri/generate-tauri-update-file.js
Comment (BrunoBernardino) [reply packages/tauri/generate-tauri-update-file.js:39]: The main advantage here is that I can reuse the objects (like `${tauriUpdate.platforms.linux.url}.sig`), otherwise we'll have to duplicate the repo URL. Both are fine, it's mostly so we have a "template" to start with that we populate, which I think makes things clearer. I'm fine to do it either way, here. So, your call.

File: packages/tauri/src-tauri/Cargo.toml
Comment (BrunoBernardino) [reply packages/tauri/src-tauri/Cargo.toml:3]: I have found that to _not_ be the case for linux, for some reason, only for macos/windows. Similarly, the file name is `padloc` in Linux, but `Padloc` in macos/windows, which is strange, but not a deal-breaker, since the app is still `Padloc` everywhere.

File: packages/tauri/src-tauri/Cargo.toml
Comment (BrunoBernardino) [reply packages/tauri/src-tauri/Cargo.toml:2]: I considered it, but it makes the menu/window name in macOS show lowercase, which is inconsistent with their guidelines. I prefer to see that warning instead, but can change it!

File: .github/workflows/release-tauri.yml
Comment (MaKleSoft) [reply .github/workflows/release-tauri.yml:13]: I think I'd like to try automatically updating the files in the repo, simply because that'll safe us more step and we can be absolutely sure the version is consistent everywhere. Can you think of any reason this may cause problems?

File: packages/tauri/src-tauri/tauri.conf.json
Comment (MaKleSoft) [reply packages/tauri/src-tauri/tauri.conf.json:61]: I definitely want to have everything in a single release, as discussed above. Previously, the separate Tauri release was really just a way to make it very clear that the Tauri app was still super experimental. Going forward it should be released alongside everything else and eventually replace the Electron version altogether.

File: packages/tauri/generate-tauri-update-file.js
Comment (MaKleSoft) [reply packages/tauri/generate-tauri-update-file.js:36]: If you're worried about repetition, you could simply store the "base url" in a variable. For me, the "canonical form" of the whole thing would look something like this:

```js
const baseUrl = `https://github.com/padloc/padloc/download/v${version}/`
const tauriUpdate = {
    name: version,
    pub_date: new Date(),
    platforms: {
        darwin: {
            url: `${baseUrl}/Padloc.app.tar.gz`,
            signature: `${baseUrl}/Padloc.app.tar.gz.sig`,
        },
        linux: {
            url: `${baseUrl}/padloc_${version}_amd64.AppImage.tar.gz`,
            signature: `${baseUrl}/padloc_${version}_amd64.AppImage.tar.gz.sig`,
        },
        win64: {
            url: `${baseUrl}/Padloc_${version}_x64.msi.zip`,
            signature: `${baseUrl}/Padloc_${version}_x64.msi.zip.sig`,
        },
    },
};
```

But that's just my preference and I feel like we're getting dangerously close to shedding bikes again (especially since this is basically just a "support script"), so I'm totally fine with keeping it as-is 😊

File: packages/tauri/src-tauri/Cargo.toml
Comment (MaKleSoft) [reply packages/tauri/src-tauri/Cargo.toml:2]: I think that's only if you're running it in dev mode. Once bundled the name in the app menu should be uppercase. Can you verify that?

**Review by MaKleSoft — APPROVED** [review]
This will probably need some tweaking in the future but I don't see any critical issues right now, and this is definitely a huge step in the right direction. Feel free to merge at your discretion!

File: packages/tauri/src-tauri/Cargo.toml
Comment (BrunoBernardino) [reply packages/tauri/src-tauri/Cargo.toml:2]: I'll change and verify this later when I pull up my macbook again, if that's fine.

File: packages/tauri/generate-tauri-update-file.js
Comment (BrunoBernardino) [reply packages/tauri/generate-tauri-update-file.js:36]: I'll do it your way (except for `new Date()`, because I want to make sure it's clear what the expectation is, and if the object would be passed around before "stringified", it could be an `int` instead of `string`, and I don't like to promote that JS inconsistency).

File: packages/tauri/src-tauri/tauri.conf.json
Comment (BrunoBernardino) [reply packages/tauri/src-tauri/tauri.conf.json:61]: Ok. Since we already have a v3 release going and I don't want to create confusion, we'll continue with the `tauri-` prefix here for now. We'll remove the prefix once this script becomes a wider release, if that makes sense.

File: .github/workflows/release-tauri.yml
Comment (BrunoBernardino) [reply .github/workflows/release-tauri.yml:13]: Yes, it can more easily cause conflicts, and, for example, I can't test this without "adding noise" to the branch (though I can then force-push without that commit), but it shouldn't be a big problem, and I can appreciate the consistency ask.

I would, however, prefer to do it in a follow-up PR, once this script becomes a release for everything, if that's fine.

**Pull request merged**
